src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt

```diff
@@ -222,26 +254,28 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
         }
         val response = CosmasProto.FileVersionListResponse.newBuilder()
         var curFileId = request.fileId
+        var rest = request.count
+        if (rest == 0) { // default value
+            rest = -1
+        }
         val versionList = mutableListOf<CosmasProto.FileVersionInfo>()
-        while (curFileId != null) {
-            val blobs: Page<Blob> = try {
-                this.storage.list(bucketName(request.info.isFreePlan), Storage.BlobListOption.versions(true),
-                        Storage.BlobListOption.prefix(fileStorageName(curFileId, request.info)))
+        while (curFileId != null && rest != 0) {
+            val fileIdVersionsInfo = try {
+                getFileIdVersionsInfo(request.info, curFileId)
             } catch (e: StorageException) {
                 handleStorageException(e, responseObserver)
                 return
             }
-            blobs.iterateAll().forEach {
-                val fileVersion = FileVersion.parseFrom(it.getContent())
-                val versionInfo = CosmasProto.FileVersionInfo.newBuilder()
-                        .setGeneration(it.generation)
-                        .setTimestamp(fileVersion.timestamp)
-                        .setFileId(curFileId)
-                versionList.add(versionInfo.build())
+            if (rest == -1) {
+                versionList.addAll(fileIdVersionsInfo.reversed())
+            } else {
+                val delta = min(rest, fileIdVersionsInfo.size)
+                versionList.addAll(fileIdVersionsInfo.reversed()
```

Comment:
Why don't you reverse the list at the very beginning? Or even better, why not store it already reversed?